packages/sources/token-balance/src/transport/evm.ts

```diff
@@ -0,0 +1,244 @@
+  ): Promise<AdapterResponse<BaseEndpointTypes['Response']>> {
+    const { addresses } = param
+
+    // verify providers exist for this request
+    const normalizedAddresses = this.verifyProviders(addresses)
+
+    // construct list of decimals RPC requests
+    const constructDecimalsKey = (network: string, contractAddress: string): string =>
+      `${network}_${contractAddress}`
+
+    const decimalsMap = new Map<string, number | null>()
+    const decimalsRequests = []
+    for (const address of normalizedAddresses) {
+      const { network, contractAddress, decimalsSignature, provider } = address
+
+      // since we can have multiple addresses with balances of the same token,
+      // skip if we've encountered this pair already
+      const decimalsKey = constructDecimalsKey(network, contractAddress)
+      if (decimalsMap.has(decimalsKey)) {
+        logger.debug(`skipping decimals fetch for contract address ${address}`)
+        continue
+      }
+      decimalsMap.set(decimalsKey, null)
+
+      // get decimals
+      const iface = new ethers.Interface([decimalsSignature])
+      const decimalsFunctionName = iface.getFunctionName(decimalsSignature)
+      const decimalsEncoded = iface.encodeFunctionData(decimalsFunctionName)
+
+      // push with callback including network and contract address
+      decimalsRequests.push(
+        provider
+          .call({
+            to: contractAddress,
+            data: decimalsEncoded,
+          })
+          .then((result) => ({
+            network,
+            contractAddress,
+            decimals: Number(iface.decodeFunctionResult(decimalsSignature, result)[0]),
+          })),
+      )
+    }
+
+    const decimalsResponses = await Promise.all(decimalsRequests)
+
+    decimalsResponses.forEach((response) => {
+      const decimalsKey = constructDecimalsKey(response.network, response.contractAddress)
+      decimalsMap.set(decimalsKey, response.decimals)
+    })
+
+    // construct list of RPC balance requests
+    const balanceRequests = []
+    for (const address of normalizedAddresses) {
+      const { network, contractAddress, wallets, balanceOfSignature, provider } = address
+      const iface = new ethers.Interface([balanceOfSignature])
+      const balanceOfFunctionName = iface.getFunctionName(balanceOfSignature)
+
+      for (const wallet of wallets) {
+        const balanceOfEncoded = iface.encodeFunctionData(balanceOfFunctionName, [wallet])
+        balanceRequests.push(
+          provider
+            .call({
+              to: contractAddress,
+              data: balanceOfEncoded,
+            })
+            .then((result) => ({
+              network: network.toLowerCase(), // return lowercase
+              contractAddress,
+              walletAddress: wallet,
+              balance: String(iface.decodeFunctionResult(balanceOfSignature, result)[0]),
+              decimals: decimalsMap.get(constructDecimalsKey(network, contractAddress)) || 0,
```

Comment:
Should throw error here instead of assuming 0

Reply:
good catch